Source/WebCore/svg/SVGLengthValue.cpp

```diff
@@ -260,45 +219,62 @@ SVGLengthValue SVGLengthValue::blend(const SVGLengthValue& from, const SVGLength
     return { WebCore::blend(fromValue.releaseReturnValue(), toValue, { progress }), to.lengthType() };
 }
 
-SVGLengthValue SVGLengthValue::fromCSSPrimitiveValue(const CSSPrimitiveValue& value, const CSSToLengthConversionData& conversionData, ShouldConvertNumberToPxLength shouldConvertNumberToPxLength)
+ExceptionOr<void> SVGLengthValue::setValueAsString(StringView valueAsString, SVGLengthMode lengthMode)
 {
-    auto primitiveType = value.primitiveType();
-    if (primitiveType == CSSUnitType::CSS_NUMBER && shouldConvertNumberToPxLength == ShouldConvertNumberToPxLength::Yes)
-        return { value.resolveAsNumber<float>(conversionData), SVGLengthType::Pixels };
-
-    switch (primitiveTypeToLengthType(primitiveType)) {
-    case SVGLengthType::Unknown:
-        return { };
-    case SVGLengthType::Number:
-        return { value.resolveAsNumber<float>(conversionData), SVGLengthType::Number };
-    case SVGLengthType::Percentage:
-        return { value.resolveAsPercentage<float>(conversionData), SVGLengthType::Percentage };
-    default:
-        return { value.resolveAsLength<float>(conversionData), SVGLengthType::Pixels };
-    }
-
-    ASSERT_NOT_REACHED();
-    return { };
+    m_lengthMode = lengthMode;
+    return setValueAsString(valueAsString);
 }
 
-Ref<CSSPrimitiveValue> SVGLengthValue::toCSSPrimitiveValue(const Element* element) const
+SVGLengthType SVGLengthValue::lengthType() const
 {
-    if (RefPtr svgElement = dynamicDowncast<SVGElement>(element)) {
-        SVGLengthContext context { svgElement.get() };
-        auto computedValue = context.convertValueToUserUnits(valueInSpecifiedUnits(), lengthType(), lengthMode());
-        if (!computedValue.hasException())
-            return CSSPrimitiveValue::create(computedValue.releaseReturnValue(), CSSUnitType::CSS_PX);
-    }
+    return WTF::switchOn(m_value,
+        [](const CSS::Number<>&) -> SVGLengthType {
+            return SVGLengthType::Number;
+        },
+        [](const CSS::LengthPercentage<>& length) -> SVGLengthType {
+            if (auto raw = length.raw())
+                return primitiveTypeToLengthType(toCSSUnitType(raw->unit));
+
+            return SVGLengthType::Unknown;
+        }
+    );
+}
 
-    return CSSPrimitiveValue::create(valueInSpecifiedUnits(), lengthTypeToPrimitiveType(lengthType()));
+bool SVGLengthValue::isZero() const
+{
+    return WTF::switchOn(m_value,
+        [](const auto& number) {
+            return number.isKnownZero();
+        },
+        [](const CSS::LengthPercentage<>& length) {
+            return length.isKnownZero();
+        }
+    );
```

Comment:
can probably make this less verbose with `const auto& value` and in many places. not sure what's preferred here? 

fwiw, auto deduced in lambdas like this is bad for type hints.

@weinig